File: entity.js
export class Entity {
   constructor({x = 0, y = 0, dx = 0, dy = 0, angle = 0, dAngle = 0,
                radius = 0, mass = 0, health = 1, damage = 0}) {
      this.x = x
      this.y = y
      this.dx = dx
      this.dy = dy
      this.angle = angle
      this.dAngle = dAngle

      this.radius = radius
      this.mass = mass
      this.health = health
      this.damage = damage

      this.createdEntities = []
   }

   //
   // Health
   //
   isAlive() {
      return this.health > 0
   }

   hitWith(entity) {
      this.health -= entity.damage

      if (this.health <= 0) {
         this.die()
      }
   }

   die() {
      // sub-classes should override this
   }

   updatePosition(dt) {
      this.x += this.dx * dt
      this.y += this.dy * dt
      this.angle += this.dAngle * dt
   }

   update(dt) {
      this.updatePosition(dt)
   }

   draw(ctx) {
      ctx.save()

      ctx.translate(this.x, this.y)
      ctx.rotate(this.angle)

      this.drawEntity(ctx)

      ctx.restore()
   }

   createEntity(entity) {
      this.createdEntities.push(entity)
   }

   getCreatedEntities() {
      return this.createdEntities.splice(0)
   }

   //
   // Distances and angles
   //

   distanceFrom(entity) {
      return this.distanceFromPoint(entity.x, entity.y)
   }

   distanceFromPoint(x, y) {
      const cx = x - this.x
      const cy = y - this.y
      return Math.sqrt(cx*cx + cy*cy)
   }

   angleTo(actor) {
      return this.angleToPoint(actor.x, actor.y)
   }

   angleToPoint(x, y) {
      return Math.atan2(y - this.y, x - this.x) - this.angle
   }

   getOffsetPosition(frontOffset, sideOffset) {
      const sinAngFront = Math.sin(this.angle)
      const cosAngFront = Math.cos(this.angle)
      const sinAngSide = Math.sin(this.angle + Math.PI/2)
      const cosAngSide = Math.cos(this.angle + Math.PI/2)

      const x = this.x + cosAngFront * frontOffset + cosAngSide * sideOffset
      const y = this.y + sinAngFront * frontOffset + sinAngSide * sideOffset

      return [x, y]
   }
   

   //
   // Collision
   //

   isCollidingWith(entity) {
      return this.distanceFrom(entity) < this.radius + entity.radius
   }

   timeUntilHit(other, buffer = 0) {
      // See when ships would collide if continuing at their current direction and rate of speed
      // This will return a negative value if the collision would have occurred in the past
      // See https://stackoverflow.com/questions/33140999/at-what-delta-time-will-two-objects-collide
      // (Line-Line was http://www.jeffreythompson.org/collision-detection/line-line.php)
      const cx = this.x - other.x
      const cy = this.y - other.y

      const dx1 = this.dx
      const dy1 = this.dy
      const dx2 = other.dx
      const dy2 = other.dy
      const vx = dx1 - dx2
      const vy = dy1 - dy2

      const rr = this.radius + other.radius + buffer

      const a = vx*vx + vy*vy
      const b = 2 * (cx * vx + cy * vy)
      const c = cx*cx + cy*cy - rr*rr

      const disc = b*b - 4*a*c

      // If the objects don't collide, the discriminant will be negative
      if (disc < 0) {
         return Number.POSITIVE_INFINITY
      }
      else {
         const t0 = (-b - Math.sqrt(disc)) / (2*a)

         return t0

         // if (t0 >= 0) {
         //    return t0
         // }
         // else {
         //    const t1 = (-b + Math.sqrt(disc)) / (2*a)

         //    console.log("t0 = " + t0)
         //    console.log("t1 = " + t1)
         //    console.log("\n")

         //    return t1 < 0 ? Number.POSITIVE_INFINITY : t1
         // }
      }
   }

   static handleBounce(e1, e2) {
      // See https://ericleong.me/research/circle-circle/#dynamic-circle-circle-collision
      // TODO: try something from this monster? http://www.euclideanspace.com/physics/dynamics/collision/twod/index.htm
      const diffX = e2.x - e1.x
      const diffY = e2.y - e1.y
      const distBetween = Math.sqrt(diffX * diffX + diffY * diffY)
      const normX = diffX / distBetween
      const normY = diffY / distBetween

      const p = 2 * (e1.dx * normX + e1.dy * normY - e2.dx * normX - e2.dy * normY) / (e1.mass + e2.mass)

      // console.log(`Before bounce: a1 ${a1.dx}, ${a1.dy}`)
      // console.log(`Before bounce: a2 ${a2.dx}, ${a2.dy}`)

      e1.dx -= p * e2.mass * normX
      e1.dy -= p * e2.mass * normY
      e2.dx += p * e1.mass * normX
      e2.dy += p * e1.mass * normY

      // console.log(`After bounce: a1 ${a1.dx}, ${a1.dy}`)
      // console.log(`After bounce: a2 ${a2.dx}, ${a2.dy}`)
   }
}

File: v1/src/entity.dart
import 'dart:html';
import 'dart:math';

import 'world.dart';

abstract class Entity {
  num x, y, dx, dy, angle, dAngle;
  num radius, mass, life = 1, decay, health, damage;
  bool isSolid;
  String color;

  Entity({this.x = 0, this.y = 0, this.dx = 0, this.dy = 0, this.angle = 0, this.dAngle = 0,
          this.radius = 0, this.mass = 0, this.isSolid = false, this.decay = 0,
          this.health = 1, this.damage = 0, this.color = 'black'});
  
  // TODO: Use this to separate positioning calls from setting mass/damage/etc?
  Entity spawn({num x = 0, num y = 0, num dx = 0, num dy = 0, num angle = 0, num dAngle = 0}) {
    this.x = x;
    this.y = y;
    this.dx = dx;
    this.dy = dy;
    this.angle = angle;
    this.dAngle = dAngle;

    return this;
  }

  Entity spawnParticle({required num startX, required num startY, num startSpread = 0, 
            num startDX = 0, num startDY = 0, num dirAngle = 0, num dirSpread = pi * 2, 
            num minSpeed = 0, required num maxSpeed, num maxSpin = 0,
            num minRadius = 0, required num maxRadius}) {
    Random random = new Random();
    angle = dirAngle + random.nextDouble() * dirSpread - dirSpread/2;
    dAngle = random.nextDouble() * (maxSpin / 2) - maxSpin;

    final dist = random.nextDouble() * startSpread;
    x = startX + cos(angle) * dist;
    y = startY + sin(angle) * dist;

    final speed = minSpeed + random.nextDouble() * (maxSpeed - minSpeed);
    dx = startDX + cos(angle) * speed;
    dy = startDY + sin(angle) * speed;

    radius = minRadius + random.nextDouble() * (maxRadius - minRadius);

    return this;
  }

  // NOTE: life decreases with time, health decreases when hit
  bool get isAlive => life > 0 && health > 0;
  
  void hitWith(Entity entity, World world) {
    health -= entity.damage;

    bleedFrom(entity, world);
    if (health <= 0) {
      die(world);
    }
  }

  void bleedFrom(Entity entity, World world) {}
  void die(World world) {}

  //
  // Distances and angles
  //
  num distanceFrom(Entity entity) => distanceFromPoint(entity.x, entity.y);
  num distanceFromPoint(num x, num y) => sqrt(pow(x - this.x, 2) + pow(y - this.y, 2));
  num angleFrom(Entity entity) => angleFromPoint(entity.x, entity.y);
  num angleFromPoint(num x, num y) => atan2(y - this.y, x - this.x) - angle;

  Point getOffsetPosition(num frontOffset, num sideOffset) {
    return new Point(
      x + cos(angle) * frontOffset + cos(angle + pi/2) * sideOffset,
      y + sin(angle) * frontOffset + sin(angle + pi/2) * sideOffset
    );
  }

  //
  // Collision
  //
  bool isCollidingWith(Entity entity) => distanceFrom(entity) < radius + entity.radius;

  num timeUntilHit(Entity other, {num buffer = 0}) {
    // See when ships would collide if continuing at their current direction and rate of speed
    // This will return a negative value if the collision would have occurred in the past
    // See https://stackoverflow.com/questions/33140999/at-what-delta-time-will-two-objects-collide
    // (Line-Line was http://www.jeffreythompson.org/collision-detection/line-line.php)

    final cx = x - other.x;
    final cy = y - other.y;
    final vx = dx - other.dx;
    final vy = dy - other.dy;
    final rr = radius + other.radius + buffer;

    final a = pow(vx, 2) + pow(vy, 2);
    final b = 2 * (cx * vx + cy * vy);
    final c = pow(cx, 2) + pow(cy, 2) - pow(rr, 2);

    final disc = b*b - 4*a*c;

    // If the objects don't collide, the discriminant will be negative
    // We only care about first intersection, so just return t0 (which uses -b)
    return disc < 0 ? double.infinity : (-b - sqrt(disc)) / (2*a);
  }

  static void handleBounce(Entity e1, Entity e2) {
    // See https://ericleong.me/research/circle-circle/#dynamic-circle-circle-collision
    // someday: try something from this monster? http://www.euclideanspace.com/physics/dynamics/collision/twod/index.htm
    final diffX = e2.x - e1.x;
    final diffY = e2.y - e1.y;
    final distBetween = sqrt(diffX * diffX + diffY * diffY);
    final normX = diffX / distBetween;
    final normY = diffY / distBetween;

    final p = 2 * (e1.dx * normX + e1.dy * normY - e2.dx * normX - e2.dy * normY) / (e1.mass + e2.mass);

    e1.dx -= p * e2.mass * normX;
    e1.dy -= p * e2.mass * normY;
    e2.dx += p * e1.mass * normX;
    e2.dy += p * e1.mass * normY;
   }

  //
  // Update
  //
  void updatePosition(num dt) {
    x += dx * dt;
    y += dy * dt;
    angle += dAngle * dt;
  }

  void update(num dt, World world) {
    life -= decay * dt;

    updatePosition(dt);
  }

  //
  // Draw
  //
  void drawEntity(CanvasRenderingContext2D ctx) {}

  void draw(CanvasRenderingContext2D ctx) {
    ctx.save();
    ctx.translate(x, y);
    ctx.rotate(angle);
    drawEntity(ctx);
    ctx.restore();
  }
}